File: ProductArrExceptSelf.java
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Three line explanation of solution in plain english

// Your code here along with comments explaining your approach
/*
Create a left array with the consecuritve product from left to right
then repeat the same from right to left and 
multiply both arrays
*/

class Solution {
    public int[] productExceptSelf(int[] nums) { 
        if(nums==null || nums.length ==0)
            return new int[0];
        
        int res[] = new int[nums.length];
        int product = 1;
        
        // Left side
        for(int i=0; i< nums.length; i++){
            res[i] = product;
            product = product * nums[i];
        }
        
        product = 1;
        // Right side
        for(int i=nums.length-1; i>= 0; i--){
            res[i] = res[i] * product;
            product = product * nums[i];
        }
        
        return res;
    }
}

File: productarrexceptself.py
#time complexity:o(n)
#space complexity:o(n)
#passed all cases on LeetCode: yes
#difficulty faced:-
# comments:
class Solution:
    def productExceptSelf(self, nums: 'List[int]') -> 'List[int]':
        
        #2 ptr approach
        # we run 2 pointers one from left and 1 from right 
        n = len(nums) - 1
        #these left and right are actually left product and right product
        #we store the product of left side elemets to the one we are in in left and right side element to one we are at to the right
        left,right = 1,1
        res = [1]*len(nums)
        
        #i and n-i are are two array indexes 
        for i in range(0,len(nums)):
            res[i] *= left
            res[n-i] *=  right
            left *= nums[i]
            right *= nums[n-i]
            
        return res